OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs

```diff
@@ -264,18 +270,41 @@ public override bool HandleKeyPress(KeyInput e)
 			return false;
 		}
 
-		static IEnumerable<Actor> SelectActorsInBoxWithDeadzone(World world, int2 a, int2 b, Func<Actor, bool> cond)
+		static IEnumerable<Actor> SelectActorsOnScreen(World world, WorldRenderer wr, IEnumerable<string> sc, Player player)
 		{
-			if (a == b || (a - b).Length > Game.Settings.Game.SelectionDeadzone)
-				return SelectActorsInBox(world, a, b, cond);
-			else
-				return SelectActorsInBox(world, b, b, cond);
+			return SelectActorsByPlayerByClass(world.ScreenMap.ActorsInBox(wr.Viewport.TopLeft, wr.Viewport.BottomRight), sc, player);
 		}
 
-		static IEnumerable<Actor> SelectActorsInBox(World world, int2 a, int2 b, Func<Actor, bool> cond)
+		static IEnumerable<Actor> SelectActorsInWorld(World world, IEnumerable<string> sc, Player player)
 		{
+			return SelectActorsByPlayerByClass(world.ActorMap.ActorsInWorld(), sc, player);
+		}
+
+		static IEnumerable<Actor> SelectActorsByPlayerByClass(IEnumerable<Actor> sel, IEnumerable<string> sc, Player player)
+		{
+			return sel.Where(a =>
+			{
+				if (a.Owner != player)
+					return false;
+				var s = a.TraitOrDefault<Selectable>();
+
+				// sc == null means that units, that meet all other criteria, get selected
+				return s != null && s.Info.Selectable && (sc == null || sc.Contains(s.Class));
+			});
+		}
```

Comment:
I can probably change something with the above three functions, but too tired right now to think clearly.